Support extended color animation interpolation

weinig opened the pull request:
#### 1140e61c63402f8b7f84b620960a1a28a2ce4513
<pre>
Support extended color animation interpolation
<a href="https://bugs.webkit.org/show_bug.cgi?id=245645">https://bugs.webkit.org/show_bug.cgi?id=245645</a>
&lt;rdar://100437917&gt;

Reviewed by Dean Jackson and Antti Koivisto.

Adds support for animating non-sRGB colors by using the new default interpolation
method, OKLab, for cases where one or more of the two keyframes in an animation
interpolation are specified using a non-legacy color syntax (e.g. color(), lch(),
etc.).

For now, there is no syntax for specifying a specific interpolation method to use,
as none has been specified yet, so only the two default spaces, sRGB for legacy
and OKLab for non-legacy are available.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/animation/color-interpolation-expected.txt:
Update results for additional passing test cases.

* Source/WebCore/platform/graphics/ColorBlending.cpp:
(WebCore::requiresLegacyInterpolationRules): Helper to choose which
kind of interpolation to use.
(WebCore::blend): When using the legacy path, also continue to clamp
to the 8-bit per channel sRGB to avoid any change for existing content.

* Source/WebCore/platform/graphics/ColorNormalization.h:
(WebCore::makeColorTypeByNormalizingComponents):
(WebCore::makeColorTypeByNormalizingComponents&lt;HWBA&lt;float&gt;&gt;):
Interpolation can produce negative and otherwise out of range
values, so ensure we clamp here. Fixes test case where the
progress was set to -0.3.

Canonical link: <a href="https://commits.webkit.org/254960@main">https://commits.webkit.org/254960@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ed6de0be8c8eddd88e6597d68b72e219f62c778

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35259 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99994 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158316 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33757 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28892 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83037 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/96393 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26871 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77502 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26712 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81647 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/81416 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69740 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/children-changed, /TestWTF:WTF_WordLock.ManyContendedLongSections (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34849 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15453 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32661 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16433 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3458 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/39371 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38348 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35522 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/platform/graphics/ColorBlending.cpp:
```diff
@@ -91,41 +91,63 @@ Color blendWithWhite(const Color& color)
 
+static bool requiresLegacyInterpolationRules(const Color& color)
+{
+    return color.callOnUnderlyingType([&] (const auto& underlyingColor) {
+        using ColorType = std::decay_t<decltype(underlyingColor)>;
+
+        if constexpr (std::is_same_v<ColorType, SRGBA<uint8_t>>)
+            return true;
+        else if constexpr (std::is_same_v<ColorType, SRGBA<float>>)
+            return true;
+        else if constexpr (std::is_same_v<ColorType, HSLA<float>>)
+            return true;
+        else if constexpr (std::is_same_v<ColorType, HWBA<float>>)
+            return true;
+        else if constexpr (std::is_same_v<ColorType, ExtendedSRGBA<float>>)
+            return !color.usesColorFunctionSerialization();
+        else
+            return false;
+    });
```
Couldn't this sort of thing achieved without all the ifs and std::is_same_vs with WTF::switchOn() type approach?

weinig replied:
Yeah, an alternate approach would look like this:

```c++
    return color.callOnUnderlyingType(WTF::makeVisitor(
        [] (const SRGBA<uint8_t>&) -> bool {
            return true;
        },
        [] (const SRGBA<float>&) -> bool {
            return true;
        },
        [] (const HSLA<float>&) -> bool {
            return true;
        },
        [] (const HWBA<float>&) -> bool {
            return true;
        },
        [&] (const ExtendedSRGBA<float>&) -> bool {
            return !color.usesColorFunctionSerialization();
        },
        [] (const auto&) -> bool {
            return false;
        }
    ));
```

Do you have a preference for one vs. another?

anttijk replied:
I think the latter looks nicer, requiring no decay_t<> etc.

anttijk reviewed (approved):
nice!

Commit:
Support extended color animation interpolation
https://bugs.webkit.org/show_bug.cgi?id=245645
<rdar://100437917>

Reviewed by Dean Jackson and Antti Koivisto.

Adds support for animating non-sRGB colors by using the new default interpolation
method, OKLab, for cases where one or more of the two keyframes in an animation
interpolation are specified using a non-legacy color syntax (e.g. color(), lch(),
etc.).

For now, there is no syntax for specifying a specific interpolation method to use,
as none has been specified yet, so only the two default spaces, sRGB for legacy
and OKLab for non-legacy are available.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/animation/color-interpolation-expected.txt:
Update results for additional passing test cases.

* Source/WebCore/platform/graphics/ColorBlending.cpp:
(WebCore::requiresLegacyInterpolationRules): Helper to choose which
kind of interpolation to use.
(WebCore::blend): When using the legacy path, also continue to clamp
to the 8-bit per channel sRGB to avoid any change for existing content.

* Source/WebCore/platform/graphics/ColorNormalization.h:
(WebCore::makeColorTypeByNormalizingComponents):
(WebCore::makeColorTypeByNormalizingComponents<HWBA<float>>):
Interpolation can produce negative and otherwise out of range
values, so ensure we clamp here. Fixes test case where the
progress was set to -0.3.

Canonical link: https://commits.webkit.org/254960@main
